### PR TITLE
Work Page | BE

### PR DESCRIPTION
**Pull request by engineervix**
[Link to Ticket](https://torchbox.monday.com/boards/1192293412/pulses/1347773123)

### Description of Changes Made

This adds a new WorkPage model to the project in line with the new design. 

**Additional Notes**

1. the old `WorkPage` model was renamed to `HistoricalWorkPage` in #51 
2. because of an existing [bug](https://code.djangoproject.com/ticket/23577) in Django, some errors (`django.db.utils.OperationalError`) were encountered when attempting to run `./manage.py migrate` after generating migrations for the new model. To mitigate this
    - I set `db_index=False` on the `social_image` field on `HistoricalWorkPage` (references: [django bug ticket](https://code.djangoproject.com/ticket/23577) and [this stackoverflow post](https://stackoverflow.com/questions/70970039/django-rename-field-and-create-new-one-with-the-same-name-returns-psycopg2-error))
    - I modified the generated migration file by introducing a couple of `migrations.RunSQL` operations to handle broken indices and constraints (references:  [django bug ticket](https://code.djangoproject.com/ticket/23577) and [this other stackoverflow post](https://stackoverflow.com/questions/76163944/django-db-utils-programmingerror-relation-already-exists-for-unique-together)). 

Here are the custom migration operations mentioned above

https://github.com/torchbox/torchbox.com/blob/8abefba95a315e367013a34e68572dee823366c2/tbx/work/migrations/0014_add_new_workpage_model.py#L24-L33

### How to Test

-  You should be able to create Work Pages in Wagtail Admin, at Pages > Torchbox > Work: http://127.0.0.1:8000/admin/pages/6/.
- the created work pages should be rendered, with all appropriate content & behaviours (services, related works, etc.) as per ACs (FE is covered in a separate ticket)

### Screenshots

<details>
  <summary>Wagtail Admin</summary>

image

</details>

<details>
  <summary>Template</summary>

image

</details>

### MR Checklist

- [x] Add a description of your pull request and instructions for the reviewer to verify your work.
- [x] If your pull request is for a specific ticket, link to it in the description.
- [x] Stay on point and keep it small so the merge request can be easily reviewed.
- [x] Tests and linting passes.

#### Unit tests

- [ ] Added
- [ ] Not required

#### Documentation

- [ ] Updated build docs
- [ ] Updated editor guidelines (https://docs.google.com/document/d/1PAWccdQ4tfaZsrEWmpDhvP3GH5RRmBOARFVp4b-kje8/edit?usp=sharing)
- [x] Not required

#### Browser testing

- [x] I have tested in the following browsers and environments (edit the list as required)
  - Latest version of Firefox on Linux
  - Latest version of Chrome on Linux
  
#### Data protection

- [x] Not relevant
- [ ] This adds new sources of PII and documents it and modifies Birdbath processors accordingly

#### Accessibility

- [ ] Automated WCAG 2.1 tests pass
- [ ] Manual WCAG 2.1 tests completed
- [ ] I have tested in a screen reader
- [ ] I have tested in high-contrast mode
- [ ] Any animations removed for prefers-reduced-motion
- [x] Not required

#### Sustainability

- [ ] Images are optimised and lazy-loading used where appropriate
- [ ] SVGs have been optimised
- [ ] Perfomance and transfer of data considered
- [ ] If JavaScript is needed alternatives have been considered
- [ ] New font variants have not been added
- [x] Not required

#### Pattern library

- [ ] The pattern library component for this template displays correctly, and does not break parent templates
- [ ] The styleguide is updated if relevant
- [ ] Changes are not relevant the pattern library
- [x] Pattern library to be updated in separate FE ticket

### CODE REVIEW

File: tbx/work/models.py [hunk 1]
```diff
@@ -143,12 +157,131 @@ def type(self):
...
+
+    @cached_property
+    def services(self):
+        return self.related_services.all()
+
+    @property
+    def first_author(self):
+        """Safely return the first author if one exists."""
+        author = self.authors.first()
+        if author:
```
Comment (ababic) [inline tbx/work/models.py:222]:
```suggestion
        if author:= self.authors.first()
```

File: tbx/work/models.py [hunk 2]
```diff
@@ -143,12 +157,131 @@ def type(self):
...
+                "read_time": case_study.read_time,
+                "related_services": case_study.related_services.all(),
+                "listing_image": case_study.header_image,
+            }
+            # get 3 pages with same services and exclude self page
+            for case_study in WorkPage.objects.filter(
+                related_services__in=self.services
+            )
+            .live()
+            .prefetch_related("related_services")
```
Comment (ababic) [inline tbx/work/models.py:255]:
```suggestion
            .prefetch_related("related_services", "authors", "authors__author")
```

File: tbx/work/models.py [hunk 3]
```diff
@@ -143,12 +157,131 @@ def type(self):
...
+        ]
+
+    def set_body_word_count(self):
+        body_basic_html = self.body.stream_block.render_basic(self.body)
+        body_text = BeautifulSoup(body_basic_html, "html.parser").get_text()
+        remove_chars = string.punctuation + "“”’"
+        body_words = body_text.translate(
+            body_text.maketrans(dict.fromkeys(remove_chars))
+        ).split()
+        self.body_word_count = len(body_words)
```
Comment (ababic) [inline tbx/work/models.py:269]: Is this a pre-existing pattern copied from somewhere else?

File: tbx/project_styleguide/templates/patterns/pages/work/work_page.html [hunk 4]
```diff
@@ -0,0 +1,108 @@
+{% extends "patterns/base_page.html" %}
+{% load wagtailcore_tags wagtailimages_tags static %}
+
+{% block meta_tags %}
+    <script>
+        window.dataLayer = window.dataLayer || [];
+        window.dataLayer.push({
+            'author-title' : '{{ page.authors.all.0.author.role }}',
```
Comment (ababic) [inline tbx/project_styleguide/templates/patterns/pages/work/work_page.html:8]:
```suggestion
            'author-title' : '{{ page.first_author.role }}',
```

File: tbx/project_styleguide/templates/patterns/pages/work/work_page.html [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            'author-name' : '{{ page.authors.all.0.author.name }}',
```
Comment (ababic) [inline tbx/project_styleguide/templates/patterns/pages/work/work_page.html:9]:
```suggestion
            'author-name' : '{{ page.first_author.name }}',
```

File: tbx/project_styleguide/templates/patterns/pages/work/work_page.html [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            'tags' : [{% for tag in page.related_services.all %}'{{tag.name}}'{% if not forloop.last %}, {% endif %}{% endfor %}],
+            'read-time' : '{{ page.read_time }}',
+            'id': '{{ page.id }}',
+            'page-h1' : '{{ page.title }}',
+            'page-title' : '{{ page.title }} | Torchbox',
+            'page-path': '{{ page.slug }}',
+            'event': 'page-information'
+        });
+    </script>
+    <meta name="author" content="{{ page.authors.all.0.author.name }}">
```
Comment (ababic) [inline tbx/project_styleguide/templates/patterns/pages/work/work_page.html:20]:
```suggestion
    <meta name="author" content="{{ page.first_author.name }}">
```

File: tbx/work/models.py [hunk 7]
```diff
@@ -143,12 +157,131 @@ def type(self):
...
+
+    @property
+    def type(self):
+        return "CASE STUDY"
+
+    @cached_property
+    def services(self):
+        return self.related_services.all()
+
+    @property
```
Comment (ababic) [inline tbx/work/models.py:218]: Let's make this value reusable once the query has been made:

```suggestion
    @cached_property
```

File: tbx/project_styleguide/templates/patterns/pages/work/work_page.html [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        {% endfor %}
+    </div>
+
+    <div class="streamfield__title">
+        <h1 class="heading heading--one">{{ page.title }}</h1>
+
+        <div class="tags tags--with-title">
+            <span class="sr-only">Filter case studies by category</span>
+            {% for tag in page.services %}
+                {% include "patterns/atoms/tag/tag.html" with tag_slug=tag.slug tag_name=tag.name tag_link_base=work_index_url %}
```
Comment (ababic) [inline tbx/project_styleguide/templates/patterns/pages/work/work_page.html:41]: If we always want to target the parent page, this should work. We might even be able to ditch `WorkPage.work_index` and `work_index_url` too, perhaps?

```suggestion
                {% include "patterns/atoms/tag/tag.html" with tag_slug=tag.slug tag_name=tag.name tag_link_base='../' %}
```

File: tbx/project_styleguide/templates/patterns/pages/work/work_page.html [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    {% if page.related_works %}
+        {% include "patterns/atoms/section-title/section-title.html" with title="More" aria_label="Related blog posts" classes="section-title--related-posts" %}
+
+        <ul class="streamfield__related-posts">
+            {% for post in page.related_works %}
+                {% include "patterns/molecules/listing/listing--image.html" with title=post.title client=post.client name=post.author.name job_title=post.author.role link=post.url date=post.date reading_time=post.read_time tags=post.related_services listing_image=post.listing_image listing_image=post.listing_image tags_screenreader_text="Filter work posts by category" %}
+            {% endfor %}
+
+            <li class="mx-auto lg:mx-0 mt-spacerSmall">
+                {% include "patterns/atoms/see-more/see-more.html" with link=work_index_url text="More case studies" %}
```
Comment (ababic) [inline tbx/project_styleguide/templates/patterns/pages/work/work_page.html:103]:
```suggestion
                {% include "patterns/atoms/see-more/see-more.html" with link='../' text="More case studies" %}
```

File: tbx/work/models.py [hunk 10]
```diff
@@ -143,12 +157,131 @@ def type(self):
...
+
+    header_image = models.ForeignKey(
+        "images.CustomImage",
+        verbose_name="Image",
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL,
+        related_name="+",
+    )
+    header_caption = models.CharField("Caption", max_length=255, blank=True)
```
Comment (ababic) [inline tbx/work/models.py:177]: Just to clarify: `verbose_name` values for model fields (and in Model.Meta) should always be lower-case, except for entity names (e.g. Google) and accronyms (e.g. NASA). Django (and by extension, Wagtail) will upper-case the first letter automatically wherever it counts.

Comment (ababic) [reply tbx/work/models.py:177]: The above is not true for `label` values on StreamField blocks or form fields, though :)

File: tbx/project_styleguide/templates/patterns/molecules/streamfield/blocks/work_section_block.html [hunk 11]
```diff
@@ -0,0 +1,10 @@
+{% load wagtailcore_tags %}
+
+<div>
+    {{ value.title }}
+</div>
+<div>
+    {% for block in value.content %}
+        {% include_block block with unique_id=block.id %}
+    {% endfor %}
```
Comment (ababic) [inline tbx/project_styleguide/templates/patterns/molecules/streamfield/blocks/work_section_block.html:9]: I feel like we should probably be doing this sort of thing in the main streamfield block template, for consistency. That way, we could always just do:
```suggestion
     {% include_block value.content %}
```

File: tbx/work/models.py (hunk 3 above)
Comment (engineervix) [reply tbx/work/models.py:269]: Yes this is what's been used on the previous build. It's also used on the `BlogPage`

File: tbx/work/models.py (hunk 1 above)
Comment (engineervix) [reply tbx/work/models.py]: Thanks! Updated in 44f8861

File: tbx/work/models.py [hunk 12]
```diff
@@ -143,12 +157,131 @@ def type(self):
...
+    @property
+    def work_index(self):
+        ancestor = WorkIndexPage.objects.ancestor_of(self).order_by("-depth").first()
+
+        if ancestor:
+            return ancestor
+        else:
+            # No ancestors are work indexes,
+            # just return first work index in database
+            return WorkIndexPage.objects.live().public().first()
```
[the remark this answers is not in the document]
Comment (engineervix) [reply tbx/work/models.py]: Good point! I ended up removing this property altogether (in 44f8861), based on your comment [here](https://github.com/torchbox/torchbox.com/pull/55#discussion_r1466728475)

File: tbx/work/models.py (hunk 2 above)
Comment (engineervix) [reply tbx/work/models.py]: Updated in 44f8861

File: tbx/project_styleguide/templates/patterns/pages/work/work_page.html (hunk 4 above)
Comment (engineervix) [reply tbx/project_styleguide/templates/patterns/pages/work/work_page.html]: Thanks, updated in 3d090a5

File: tbx/project_styleguide/templates/patterns/pages/work/work_page.html (hunk 6 above)
Comment (engineervix) [reply tbx/project_styleguide/templates/patterns/pages/work/work_page.html]: Updated in 3d090a5

File: tbx/project_styleguide/templates/patterns/pages/work/work_page.html (hunk 9 above)
Comment (engineervix) [reply tbx/project_styleguide/templates/patterns/pages/work/work_page.html]: Thanks, updated in 3d090a5

File: tbx/work/models.py (hunk 7 above)
Comment (engineervix) [reply tbx/work/models.py]: done in 44f8861

File: tbx/project_styleguide/templates/patterns/pages/work/work_page.html (hunk 5 above)
Comment (engineervix) [reply tbx/project_styleguide/templates/patterns/pages/work/work_page.html]: Thank you! Done in 3d090a5

**Pull request merged**